clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/ingestion/heartbeat/HeartbeatVersionedStats.java

```diff
@@ -23,4 +34,42 @@ public void recordLeaderLag(String storeName, int version, String region, long l
   public void recordFollowerLag(String storeName, int version, String region, long lag) {
     getStats(storeName, version).recordFollowerLag(region, lag);
   }
+
+  @Override
+  public synchronized void loadAllStats() {
+    Set<String> stores = getReportedStores();
+    metadataRepository.getAllStores().forEach(store -> {
+      if (stores.contains(store.getName())) {
+        addStore(store.getName());
+        updateStatsVersionInfo(store.getName(), store.getVersions(), store.getCurrentVersion());
+      }
+    });
+  }
+
+  @Override
+  public void handleStoreCreated(Store store) {
+    Set<String> stores = getReportedStores();
+    if (stores.contains(store.getName())) {
+      addStore(store.getName());
+    }
+  }
+
+  @Override
+  public void handleStoreChanged(Store store) {
+    Set<String> stores = getReportedStores();
+    if (stores.contains(store.getName())) {
+      updateStatsVersionInfo(store.getName(), store.getVersions(), store.getCurrentVersion());
+    }
+  }
+
+  Set<String> getReportedStores() {
```

Comment:
I am trying to understand why we need to do this. Is it because `metadataRepository` could have some stale stores that are no longer exist on the host?

Reply:
No.  The reason why we have to do this is far dumber then that.  The parent class of this class calls it's own methods from it's constructor.  Specifically, it calls the loadAllStats() function which in turn invokes our own overridden method and then calls this.  Since in java the first line of constructor for a subclass has to be super(...), this means that this function will get called before these maps are assigned/initialized.  So we have to add this dumb defensive check for that.

The TODO is here to tell us that when we get around to fixing that code, we can remove this.

Reply:
Actually, I think I misunderstood your question.  That is right.  As partitions are moved off of hosts, we'd like to stop reporting on them.